src/python/WMCore/MicroService/MSPileup/MSPileupTasks.py

```diff
@@ -100,6 +104,95 @@ def pileupSizeTask(self):
             self.logger.exception(msg)
         return cmsDict, cusDict
 
+    def partialPileupTask(self, rseExpression):
+        """
+        Execute partial pileup placement according to the following logic:
+
+        - get rucio DIDs (datasets, in CMS name blocks) for our pileup document
+        - get portion of DIDs based on ceil(containerFraction * num_rucio_datasets)
+        - create new container DID in Rucio via Rucio wrapper createContainer API
+          which calls rucio CLI add_container
+        - update MSPileup JSON with
+          - create new custom Name as pileup+extention
+          - add new transition record and update MSPileup document
+        - we call attachDIDs Rucio wrapper API with our set of DIDs and rses from pileup document
+        - create new rules for custom DID via provided rseExpression
+        - set expiration date (to be 24h) for already existing ruleIds from pileup document
+        - add new ruleIds to pileup document
+
+        :param rseExpression: string, Rucio rse expression, see
+        https://rucio.cern.ch/documentation/started/concepts/rse_expressions/
+        :return: None
+        """
+        spec = {'containerFraction': {"$gt": 0}}
+        docs = self.mgr.getPileup(spec)
+        for doc in docs:
+            self.logger.info("partialPileupTask process %s", doc)
+            fraction = doc['containerFraction']
+
+            # check previous fraction value and proceed only if has changed
+            lastFraction = 0
+            transition = doc.get('transition', [])
+            if transition:
+                lastFraction = transition[-1]['containerFraction']
+            if lastFraction == fraction:
+                # do nothing since there is no transition, i.e. skip this record
+                continue
+
+            # get rucio DIDs (datasets, in CMS name blocks) for our pileup document as following:
+            # - when container fraction is decreasing: new container is a subset of customName
+            # - when container fraction is increasing: new container is a superset of customName + subset of pileupName
+
+            # by default we assume it is decreasing
+            pname = doc['customName']
+            blockNames = self.rucioClient.getBlocksInContainer(pname)
+
+            # but if it increasing
+            if lastFraction < fraction:
```

Comment:
no, everything is correct here. The `lastFraction` is last fraction. And `fraction` is current one. Therefore to have increasing direction we need `lastFraction < fraction` condition.

Reply:
From what you said, my understand is:
* `lastFraction` is the desired fraction to have on Disk
* while `fraction` is the current fraction on Disk (which will change)

Based on that, I have to disagree with your interpretation. If a pileup has fraction=0.5 and lastFraction=0.25, then that means that the user wants to decrease the data availability on Disk, by another quarter of the original pileup container. Hence, the data availability on Disk is actually decreasing.

Reply:
Valentin wrote: "The fraction and lastFraction has no definition if it is related to Disk or not. Both of them are containterFraction which we defined as "containerFraction": real number with the container fraction to be distributed (optional). Is it Disk or Tape fraction is not clear to me and requires further clarification. If you need such distinction I suggest to address it in different issue/PR as this one is too complicated already."

Reply:
No, there is no need to couple Disk or Tape to these fraction variables. I used those words only for explanation purposes.

Rephrasing that explanation:
* fraction is the current pileup data availability;
* while lastFraction is the (future) desired pileup data availability.

hence, if fraction is greater than lastFraction, that means that the user wants to shrink pileup data presence.
If fraction is smaller than lastFraction, that means the user wants to have more pileup blocks available.

Please let me know if things are still not clear and I can explain it over Zoom (as I still have a good internet access for another hour or so ;)

Reply:
Alan, why you insist that `lastFraction` is the (future)? It is comes from last (previous) transition record, i.e. it is in a past not in a future. Here is how it is defined:
```
            # check previous fraction value and proceed only if has changed
            lastFraction = 1.0
            transition = doc.get('transition', [])
            if transition:
                lastFraction = transition[-1]['containerFraction']
            if lastFraction == fraction:
                # do nothing since there is no transition, i.e. skip this record
                self.logger.info("Pileup name %s has no container fractions changes", doc['pileupName'])
                continue

            # get rucio DIDs (datasets, in CMS name blocks) for our pileup document as following:
            # - when container fraction is decreasing: new container is a subset of customName
            # - when container fraction is increasing: new container is a superset of customName + subset of pileupName

            # if it increasing
            if lastFraction < fraction:
```
and even all comments are about past and not about the future. I think confusion comes from wrong value assigned initially to it. I changed it to 1 instead of 0 (as it was originally done in a code).

Reply:
Oh, thank you for clarifying it. Yes, the default of 1.0 does help.

I still think the definition of the transition record, e.g.:
```
            transitionRecord = {'containerFraction': fraction,
                                'customDID': customDID(pname),
                                'updateTime': gmtimeSeconds(),
                                'DN': userDN}
```
is and will lead to confusion, given that we are actually defining things from different actions of the service.
If you can join a Zoom room soon, please ping me on mattermost.

Reply:
In case we don't make it to Zoom, let me try to express my concerns here:
1) I'd suggest to rename `lastFraction` to `previousFraction`
2) when the transition record is created, it should keep data concerning a single state/transition/action. With the current implementation, it will be a mix up of previous state with the new state (new container fraction). Example:
a) user DN will be associated with the previous containerFraction
b) updateTime corresponds to the time that that containerFraction was overriden by a new one
c) customName is created for that transition, but actually that container fraction was using a previous pileup name.